contracts/limitOrder/LimitOrderBook.sol

```diff
@@ -0,0 +1,170 @@
+    // EXTERNAL NON-VIEW
+    //
+
+    function initialize(
+        string memory name,
+        string memory version,
+        address clearingHouseArg,
+        address limitOrderFeeVaultArg
+    ) external initializer {
+        __ReentrancyGuard_init();
+        __OwnerPausable_init();
+        __EIP712_init(name, version); // ex: "PerpCurieLimitOrder" and "1"
+
+        // LOB_CHINC : ClearingHouse Is Not Contract
+        require(clearingHouseArg.isContract(), "LOB_CHINC");
+        clearingHouse = clearingHouseArg;
+        accountBalance = IClearingHouse(clearingHouse).getAccountBalance();
+
+        // LOB_LOFVINC : LimitOrderFeeVault Is Not Contract
+        require(limitOrderFeeVaultArg.isContract(), "LOB_LOFVINC");
+        limitOrderFeeVault = limitOrderFeeVaultArg;
+    }
+
+    function setClearingHouse(address clearingHouseArg) external onlyOwner {
+        // LOB_CHINC: ClearingHouse Is Not a Contract
+        require(clearingHouseArg.isContract(), "LOB_CHINC");
+        clearingHouse = clearingHouseArg;
+        accountBalance = IClearingHouse(clearingHouse).getAccountBalance();
+
+        emit ClearingHouseChanged(clearingHouseArg);
+    }
+
+    function setLimitOrderFeeVault(address limitOrderFeeVaultArg) external onlyOwner {
+        // LOB_LOFVINC: LimitOrderFeeVault Is Not a Contract
+        require(limitOrderFeeVaultArg.isContract(), "LOB_LOFVINC");
+        limitOrderFeeVault = limitOrderFeeVaultArg;
+
+        emit LimitOrderFeeVaultChanged(limitOrderFeeVaultArg);
+    }
+
+    /// @inheritdoc ILimitOrderBook
+    function fillLimitOrder(
+        LimitOrder memory order,
+        bytes memory signature,
+        uint80 roundIdWhenTriggered
+    ) external override nonReentrant {
+        bytes32 orderHash = getOrderHash(order);
+        verifySigner(order, signature);
+
+        // TODO: support StopLimitOrder in the future
+        // LOB_OSLO: Only Support LimitOrder
+        require(order.orderType == ILimitOrderBook.OrderType.LimitOrder, "LOB_OSLO");
+
+        // LOB_OIFA: Order is filled already
+        require(_ordersStatus[orderHash] != ILimitOrderBook.OrderStatus.Filled, "LOB_OIFA");
+        // LOB_OIC: Order is cancelled
+        require(_ordersStatus[orderHash] != ILimitOrderBook.OrderStatus.Cancelled, "LOB_OIC");
+
+        int256 oldTakerPositionSize = IAccountBalance(accountBalance).getTakerPositionSize(
+            order.trader,
+            order.baseToken
+        );
+
+        (uint256 base, uint256 quote) = IClearingHouse(clearingHouse).openPositionFor(
+            order.trader,
+            IClearingHouse.OpenPositionParams({
+                baseToken: order.baseToken,
+                isBaseToQuote: order.isBaseToQuote,
+                isExactInput: order.isExactInput,
+                amount: order.amount,
+                oppositeAmountBound: order.oppositeAmountBound,
+                deadline: order.deadline,
+                sqrtPriceLimitX96: 0,
+                referralCode: ""
```

Comment:
Do we not allow referral program to use our limit order? Suggest we could support it if there is no strong preference.